README.md
<img src="https://raw.githubusercontent.com/dagraham/nts-dgraham/master/ntslogo.png" alt="nts"  width="180px" />


# Note Taking Simplified

### Overview

_nts_ is pure python and runs in a terminal window on any platform that supports python >= 3.7.3. Its purpose is to facilitate both recording and retrieving information. Further information is available at [github](https://github.com/dagraham/nts-dgraham), the _nts_ discussion group at [groups.io](https://groups.io/g/nts) and [PyPI](https://pypi.org/project/nts-dgraham/).

Notes are recorded in plain text files with the extension ".txt" located anywhere below the _nts data_ directory. Note files have a simple format:

* Each note begins with a "+" in the first column followed by a space and then the title of the note.

* Tags, if given, are comma separated and enclosed in parentheses after the note title.

* The _note body_ begins on the next line and continues until another note or the end of the file is reached.

* Lines that begin with one or more white space characters and then "+" are treated as part of the _note body_ and not as the beginning of a new note.

* White space in the _note body_ is preserved but whitespace between notes is ignored.

* Hidden files, i.e. files with names beginning with a period, are ignored.

* Directory and file names should not contain spaces.

Suppose, e.g., that the _nts_ data directory contains a single file

    ~/nts/data/parent/child/grandchild.txt

with this content:

    ---------------- grandchild.txt begins ---------------
    + note a (red, green)
        The body of note a goes here

    + note b (blue, green)
        The body of note b here

    + note c (red, blue)
        And the body of note c here
    ---------------- grandchild.txt ends -----------------

*nts* provides two main views of this data.

* _Path View_:

        └── parent 1
            └── child 2
                └── grandchild.txt 3
                        + note a (red, green) 3-1
                        + note b (blue, green) 3-2
                        + note c (red, blue) 3-3

* _Tag View_:

        ├── blue 1
        │       + note b (blue, green) 1-1
        │       + note c (red, blue) 1-2
        ├── green 2
        │       + note a (red, green) 2-1
        │       + note b (blue, green) 2-2
        └── red 3
                + note a (red, green) 3-1
                + note c (red, blue) 3-2

Both views are outlines with branches that end with leaves that correspond to note title lines. In path view, the nodes along the branches correspond to directory or file names and there can be many of these in each branch. In tag view, on the other hand, the nodes correspond to tag names and there can only be one of these in each branch.

The numeric identifiers appended to the lines in both views are provided by _nts_. These are single numbers for the *nodes* in the outline branches that have children and hyphenated numbers for the *leaves*, e.g., in path view the "3" appended to the "grandchild.txt" node and the "3-1" appended to the "+ note a (red, green)" leaf. In the leaf identifier, the first of the two numbers is the indentifier of the parent node. These identifiers are the IDENT arguments in the _Command Summary_ given below.


### Usage

*nts* provides two ways of interacting with the data.

* Command mode

    Commands are entered at the terminal prompt. E.g., enter

        $ nts -p

    to display the path view in the terminal window. The output can also be piped in the standard way, e.g.,

        $ nts -p | less


* Session mode

    Use the `-s` argument to begin session mode:

        $ nts -s

    This begins a session in which data is loaded into memory and remains available for subsequent interaction. In this mode, *nts* assumes command of the terminal window. Then, e.g., pressing `p` would display the path view. Session mode adds several features not available in command mode, e.g., scrolling, incremental search, auto-completion and auto-suggestion.

#### Command Summary

Action          | Command Mode     | Session Mode    | Notes
----------------|------------------|-----------------|------
help            |  -h              |  h              |  ~
begin session   |  -s              |  ~              |  ~
end session     |   ~              |  ^q or F8       |  ~
path view       |  -p              |  p              |  ~
tags view       |  -t              |  t              |  ~
hide leaves     |  -l              |  l              |  l
hide branches   |  -b              |  b              |  b
copy view       |   ~              |  c              |  c
set max levels  |  -m MAX          |  m MAX          |  m
search          |                  |  / SEARCH       |  /
find REGEX      |  -f [!]REGEX     |  f [!]REGEX     |  f
get REGEX       |  -g REGEX        |  g REGEX        |  g
join JOIN       |  -j JOIN         |  j JOIN         |  j
inspect IDENT   |  -i IDENT        |  i IDENT        |  i
edit IDENT      |  -e IDENT        |  e IDENT        |  e
add to IDENT    |  -a IDENT [NAME] |  a IDENT [NAME] |  a
refresh         |   ~              |  r              |  r
edit cfg.yaml   |   ~              |  y              |  y
version check   |  -v              |  v              |  v

- l: Suppress showing leaves in the outline. In session mode this toggles the display of leaves off and on.

- b: Suppress showing branches in the outline, i.e., display only the leaves. In session mode this toggles the display of the branches off and on.

- c: Copy the active view to the system clipboard.

- m: Limit the diplay of nodes in the branches to the integer MAX levels below the starting node. Use MAX = 0 to display all levels.

- /|?: Start a case-insensitive, incremental search forward (/) or backward (?) for SEARCH. When the search is active, press "n" to continue the search in the same or "N" reverse direction,  ",," (two commas successively) to clear the search or ".." to apply the search to the complete notes of the active view.

- f: Show notes in the current view whose content contains a match for the case-insensitive regex REGEX. Press ",," to clear the search highlighting.

- g: Display note titles that contain a match in the nodes leading to the note for the case-insensitive regular expression REGEX.

- j: Display note titles for notes with tags satisfying JOIN. E.g. if JOIN = "red", then notes containing the tag "RED" would be displayed. If JOIN = "| red, blue" then notes with _either_ the tag "red" _or_ the tag "blue" would be displayed. Finally, if JOIN = "& red, blue", then notes with _both_ the tags "red" _and_ "blue" would be displayed. In general JOIN = [|&] comma-separated list of case-insensitive regular expressions.

- i: If IDENT is the 2-number identifier for a note, then display the contents of that note. Else if IDENT is the identifier for a ".txt" file, then display the contents of that file. Otherwise limit the display to that part of the outline which starts from the corresponding node. Use IDENT = 0 to start from the root node.

- e: If IDENT corresponds to either a note or a ".txt" file, then open that file for editing and, in the case of a note, scroll to the beginning line of the note.

- a: If IDENT corresponds to either a note or a ".txt" file, then open that file for appending a new note. Otherwise, if IDENT corresponds to a directory and NAME is provided, add a child called NAME to that node. If NAME ends with ".txt", a new note file will be created. Otherwise, a new subdirectory called NAME will be added to the node directory. Use "0" as the IDENT to add to the root (data) node. In command mode, "IDENT NAME" should be wrapped in quotes.

- r: reload data from the files in the data directory to incorporate external changes.

- y: open cfg.yaml in the external editor and, when the editor is closed,  incorporate any modifications into the active session.

- v: Compare the installed version of nts with the latest version on GitHub (requires internet connection) and report the result.

Here is a link to a series of short videos illustrating basic usage:
[![workflow](https://raw.githubusercontent.com/dagraham/nts-dgraham/master/workflow.png "nts playlist")](https://www.youtube.com/playlist?list=PLN2WQIqrwSxx5beH7Qn8RC25xdoz-wEHY)

There are no commands in _nts_ to remove either a file or a directory. Please use your favorite file manager for these risky actions and don't forget to restart _nts_ to update its display.

Note: when your configuraton settings specify that the external editor should open in the same terminal as _nts_, it may be necessary to press "Ctrl-L" to restore the display after closing the editor.

#### Auto completion and auto suggestion

In session mode, if one of the IDENT commands (a, e or i) is invoked when the cursor is on a line ending with an IDENT, then that IDENT will automatically be entered for the command. Otherwise, when manually entering an IDENT, pressing the tab key will cycle through IDENTs that complete the current entry.

As an illustration of auto-suggestion, a common workflow is to examine a note and then, perhaps, edit it.  In the screenshot above I clicked on the word "action" in the line that ends with `4-4` as its IDENT. Pressing "i", would then prompt for the IDENT:

    > 4-4

with `4-4` already entered. Pressing "return" would then display the details for this note as:

                        ./tagsort.txt 4-4
    + assigned for action (assigned bob)
        In tag view, items with this tag will be sorted
        in a third group and, within that group, by the
        name to whom it was assigned

With the cursor still on the top line ending with the same `4-4` IDENT, pressing "e" would prompt for the IDENT:

    > 4-4

with the `4-4` already entered. Pressing "return" would then open this same note in the external editor. Note that just five actions were involved: one mouse click and four key presses: "i", "return", "e" and "return".


### Configuration

_nts_ runs in a terminal window that should be configured to use a fixed width (monospaced) font. _Menlo Regular_ for OSX and _DejaVu Sans Mono_ for Linux are good choices. A width of at least 60 characters for the terminal window is recommended.

Before you start *nts* for the first time, think about where you would like to keep your personal data files and any log files that _nts_ will create. This will be your nts _home directory_. The _nts_ configuration file, "cfg.yaml" will be placed in this directory as well as the data and log files in the subdirectories _data_ and _logs_, respectively.

The default is to use the current working directory when you start _nts_ as the _home directory_ either 1) if it is empty (unused so far) or 2) if it contains  subdirectories called "data" and "logs" or 3) if it contains a file called "cfg.yaml" and a subdirectory called "logs". To use this option just change to this directory before starting _nts_.

Alternatively, if the current working directory doesn't satisfy the requirments but there is an environmental variable, `NTSHOME`, that contains the path to an existing directory, then *nts* will use this as its _home directory_. To use this option, first create the directory and then set the enivonmental variable by, e.g., appending the following to your "~/.bash_profile":

        export NTSHOME="complete path to the nts home directory"

Finally, if neither of the previous alternatives are satisfied, then *nts* will use "\~/nts" as its _home directory_, creating this directory if necessary.

The _nts_ "data" and "logs" directories will be created if necessary as well as the _nts_ configuration file, "cfg.yaml" using default settings. If "data" needs to be created, the user will additionally be offered the opportunity to populate it with illustrative data. Here are the default contents of this file when the "dark" background option is chosen at installation:



    # Changes to this file only take effect when nts is restarted.
    # EDIT
    # The following are examples using the editor vim. Tip: to use the
    # native version of vim under Mac OSX, replace 'vim' in edit_command
    # with:
    #        /Applications/MacVim.app/Contents/MacOS/Vim
    edit_command: vim
    # session_edit_args: arguments to edit {filepath} at {linenum} in
    # session mode and await completion
    session_edit_args: -g -f +{linenum} {filepath}
    # session_add: arguments to edit {filepath} at end of file in session
    # mode and await completion
    session_add_args: -g -f + {filepath}
    # command_edit_args: arguments to edit {filepath} at {linenum} in
    # command mode without waiting
    command_edit_args:  +{linenum} {filepath}
    # command_add_args: arguments to edit {filepath} at end of file
    # in command mode without waiting
    command_add_args: + {filepath}
    # TAG SORT
    # For listed keys, sort by the corresponding value. E.g. In tag view
    # items with the tag "now" will be sorted as if they had the tag "!".
    # Replace, remove or add keys and values with whatever you like.
    tag_sort:
        now:            '!'
        next:           '#'
        assigned:       '%'
        someday:        '&'
        completed:      '('
    # STYLE
    light_background: false
    style:
        status:         '#FFFFFF bg:#396060'
        status.key:     '#FFAA00'
        not-searching:  '#888888'
        highlighted:    '#000000 bg:#FFFF75'
        plain:          '#FAFAFA bg:#1D3030'

When the "light" background option is chosen the relevant lines are changed to these:

    light_background: true
    style:
        status:         '#FFFFFF bg:#437070'
        status.key:     '#FFAA00'
        not-searching:  '#888888'
        highlighted:    '#1D3030 bg:#A1CAF1'
        plain:          '#000000 bg:#FFF8DC'


If you make changes to "cfg.yaml" and would like to restore the defaults just delete the relevant settings from the file and restart _nts_ - the missing settings will be restored with their default values.

From time to time, new versions of _nts_ may add new settings to "cfg.yaml". When this happens, the new settings will automatically be added to your "cfg.yaml" the next time you start _nts_.


### View Sorting

The default values for _tag_sort_ in "cfg.yaml" mean that notes will be sorted in in _tag_view_ so that items with the tag "now" will be sorted as if the tag were "!", items with the tag "next" as if the tag were "#" and so forth. Tags not listed in _tag_sort_ will sorted using the actual tag. The  _dictionary sorting order_ for common keyboard characters in python is:

`! # % & ( ) * + - / 1 2 3 ; < = > ? @ A B C [ ] ^ _ a b c { } ~`

_nts_ implicitly assigns the tag "~" to notes without tags and, because of this sorting order, untagged items are listed last in _Tags View_.


To illustrate this tag sorting with the default configuration, suppose the file

    ~/nts/data/tagsort.txt

is added to the grandchild example given above with the following content:


    ---------------- tagsort.txt begins ---------------
    + action required as soon as possible (now)
        In tag view, items with this tag will be sorted
        first

    + action needed when time permits (next)
        In tag view, items with this tag will be sorted
        after 'now' items

    + assigned for action (assigned joe)
        In tag view, items with this tag will be sorted
        in a third group and, within that group, by the
        name to whom it was assigned

    + assigned for action (assigned bob)
        In tag view, items with this tag will be sorted
        in a third group and, within that group, by the
        name to whom it was assigned

    + review from time to time for action (someday)
        In tag view, items with this tag will be sorted
        after all tagged items other than 'completed'

    + finished but kept for reference (completed)
        In tag view, items with this tag will be sorted
        after all tagged items

    + a note with no tags
        In tag view, such items will be sorted last under
        the implicit tag '~'
    ---------------- tagsort.txt ends -----------------

With this addition, _Path View_ appears as:

    ├── parent 1
    │   └── child 2
    │       └── grandchild.txt 3
    │               + note a (red, green) 3-1
    │               + note b (blue, green) 3-2
    │               + note c (red, blue) 3-3
    └── tagsort.txt 4
            + action required as soon as possible (now) 4-1
            + action needed when time permits (next) 4-2
            + assigned for action (assigned joe) 4-3
            + assigned for action (assigned bob) 4-4
            + review from time to time for action (someday) 4-5
            + finished but kept for reference (completed) 4-6
            + a note with no tags 4-7


Sorting in this view is dictionary order for sibling nodes but notes are listed in the order in which they occur in the file. E.g., the siblings "parent" and "tagsort.txt" are in dictionary order but the notes in each file are listed in the order in which they occur in the file.


_Tags View_ reflects the _tag_sort_ setting in "cfg.yaml" with the _Getting Things Done_ tags _now_, _next_,  _assigned_, _someday_ and _completed_ first in that order, then _blue_, _green_ and _red_ in the middle in dictionary order and finally _~_ (no tags) last:


    ├── now 1
    │       + action required as soon as possible (now) 1-1
    ├── next 2
    │       + action needed when time permits (next) 2-1
    ├── assigned bob 3
    │       + assigned for action (assigned bob) 3-1
    ├── assigned joe 4
    │       + assigned for action (assigned joe) 4-1
    ├── someday 5
    │       + review from time to time for action (someday) 5-1
    ├── completed 6
    │       + finished but kept for reference (completed) 6-1
    ├── blue 7
    │       + note b (blue, green) 7-1
    │       + note c (red, blue) 7-2
    ├── green 8
    │       + note a (red, green) 8-1
    │       + note b (blue, green) 8-2
    ├── red 9
    │       + note a (red, green) 9-1
    │       + note c (red, blue) 9-2
    └── ~ 10
            + a note with no tags 10-1

  Note also that within the _assigned_ tags, the sorting is in dictionary order with _assigned bob_ followed by _assigned joe_ even though _assigned joe_ occured before _assigned bob_ in the file.


### Installation

#### For use in a virtual environment

The steps for OS/X or linux are illustrated below. For details see [python-virtual-environments-a-primer](https://www.google.com/url?q=https%3A%2F%2Frealpython.com%2Fpython-virtual-environments-a-primer%2F&sa=D&sntz=1&usg=AFQjCNFh7QpJQ4rPCDjZ1eLrV1BRCCpSmw).

Open a terminal and begin by creating a new directory/folder for the virtual environment, say `nts-pypi`, in your home directory:

    $ mkdir ~/nts-pypi
    $ cd ~/nts-pypi

Now continue by creating the virtual environment (python >= 3.7.3 is required for nts):

    $ python3 -m venv env

After a few seconds you will have an `./env` directory. Now activate the virtual environment:

    $ source env/bin/activate

The prompt will now change to something containing `(env)` to indicate that the virtual environment is active. Updating pip is now recommended:

    (env) $ pip install -U pip

Note that this invokes `./env/bin/pip`. Once this is finished, use pip to install nts:

    (env) $ pip install -U nts-dgraham

This will install nts and all its requirements in

    ./env/lib/python3.x/sitepackages

and will also install an executable called `nts` in `./env/bin`.You can then start nts using

    (env) nts ARGS

using ARGS enumerated in the **Command Summary** section above.


#### For use system wide

If your system allows you to run `sudo` and you want general access system wide, then you could instead install nts using

    $ sudo -H python3.x -m pip install -U nts-dgraham

replacing the `3.x` with the verion of python you want to use, e.g., `3.7`. This would put nts in your path (in the bin directory for python3.7).


You can then open a terminal and start nts using

    $ nts ARGS

using ARGS enumerated in the **Command Summary** section above.

### License

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version. See [GNU General Public License](http://www.gnu.org/licenses/gpl.html) for more details. This program is distributed in the hope that it will be useful, but *without any warranty*; without even the implied warranty of *merchantability* or *fitness for a particular purpose*.



CHANGES.txt
Recent changes as of 2022-01-27T16:53:05.463951-05:00:
- 0 seconds ago (HEAD -> working, tag: 1.3.3) Daniel Graham
    a7633d7 2022-01-27 16:53:05 -0500
    Tagged version 1.3.3.

- 16 seconds ago Daniel Graham
    de4334a 2022-01-27 16:52:49 -0500
    Added option to edit the configuration file in session mode using
    the external editor and then, when done, incorporate any
    modifications into the active session.

- 25 hours ago (origin/master, master) Daniel Graham
    546e65f 2022-01-26 16:21:36 -0500
    Tagged version 1.3.2. Implemented auto-suggest and auto-complete
    for a, e and i commands in session mode.

- 25 hours ago Daniel Graham
    fcb3b26 2022-01-26 16:20:19 -0500
    Implemented auto-suggest and auto-complete for the IDENT commands
    a, e and i in session mode with documentation in README.

- 4 days ago Daniel Graham
    dcddcc3 2022-01-23 12:44:12 -0500
    README tweaks

- 4 days ago Daniel Graham
    8d49958 2022-01-23 11:39:09 -0500
    Tagged version 1.3.1.

- 4 days ago Daniel Graham
    7358c30 2022-01-23 11:38:51 -0500
    Fixed bug in setting cfg.yaml.

- 4 days ago Daniel Graham
    37d22e9 2022-01-23 11:12:42 -0500
    Tagged version 1.3.0. Changed cfg.yaml to support dark or light
    background option.

- 4 days ago Daniel Graham
    23488a1 2022-01-23 11:11:06 -0500
    Added option to choose dark or light background settings in
    initial setup or when style settings are missing. Simplified
    cfg.yaml.

- 5 days ago Daniel Graham
    fccaced 2022-01-22 21:57:19 -0500
    Added suggestions for light style settings to README. Moved
    someday to & to place it in the first group in tag view.

- 5 days ago Daniel Graham
    1c984c9 2022-01-22 15:12:29 -0500
    Implemented the etm status and textarea colors including the plain
    style for both foreground and background colors.

- 5 days ago Daniel Graham
    fef4687 2022-01-22 12:19:56 -0500
    Tagged version 1.2.5.

- 5 days ago Daniel Graham
    3568346 2022-01-22 12:18:02 -0500
    README fixes.

- 5 days ago Daniel Graham
    ae292d1 2022-01-22 11:35:24 -0500
    Cleaned up and centered 'headers'. Cleaned up comments and
    logger.debugs. Don't highlight find match in header.

- 6 days ago Daniel Graham
    0994ebb 2022-01-21 21:13:00 -0500
    Tagged version 1.2.4.

- 6 days ago Daniel Graham
    af9fd7b 2022-01-21 21:12:20 -0500
    README tweaks. Changed path for tagsort.txt and uploaded new
    'basic workflow' video to youtube with link in README.

- 6 days ago Daniel Graham
    e2efcdf 2022-01-21 10:00:50 -0500
    Tagged version 1.2.3. Implemented highlighting for find matches

- 6 days ago Daniel Graham
    0615a9d 2022-01-21 09:59:14 -0500
    Added find highlighting to command mode.

- 7 days ago Daniel Graham
    31397a6 2022-01-20 14:30:07 -0500
    Added highlighting for find and removed line markers.

- 8 days ago Daniel Graham
    5677232 2022-01-19 09:39:34 -0500
    Removed initial indent from wrap


nts/__main__.py
#tp!/usr/bin/env python3
import sys
import os
import logging
import logging.config
logging.getLogger('asyncio').setLevel(logging.WARNING)
logger = logging.getLogger()
from prompt_toolkit import prompt
from prompt_toolkit.styles import Style
from prompt_toolkit.styles.named_colors import NAMED_COLORS
from copy import deepcopy

import ruamel.yaml
yaml = ruamel.yaml.YAML()

# for cfg.yaml

def get_yaml_data(cfg_path):
    dark = """\
light_background: false
style:
    status:             '#FFFFFF bg:#396060'
    status.key:         '#FFAA00'
    not-searching:      '#888888'
    highlighted:        '#000000 bg:#FFFF75'
    plain:              '#FAFAFA bg:#1D3030'
"""

    light = """\
light_background: true
style:
    status:               '#FFFFFF bg:#437070'
    status.key:           '#FFAA00'
    not-searching:        '#888888'
    highlighted:          '#1D3030 bg:#A1CAF1'
    plain:                '#000000 bg:#FFF8DC'
"""

    default_template= """\
# Changes to this file only take effect when nts is restarted.
# EDIT
# The following are examples using the editor vim. Tip: to use the
# native version of vim under Mac OSX, replace 'vim' in edit_command
# with:
#        /Applications/MacVim.app/Contents/MacOS/Vim
edit_command: vim
# session_edit_args: arguments to edit {filepath} at {linenum} in
# session mode and await completion
session_edit_args: -g -f +{linenum} {filepath}
# session_add: arguments to edit {filepath} at end of file in session
# mode and await completion
session_add_args: -g -f + {filepath}
# command_edit_args: arguments to edit {filepath} at {linenum} in
# command mode without waiting
command_edit_args:  +{linenum} {filepath}
# command_add_args: arguments to edit {filepath} at end of file
# in command mode without waiting
command_add_args: + {filepath}
# TAG SORT
# For listed keys, sort by the corresponding value. E.g. In tag view
# items with the tag "now" will be sorted as if they had the tag "!".
# Replace, remove or add keys and values with whatever you like.
tag_sort:
    now:        '!'
    next:       '#'
    assigned:   '%'
    someday:    '&'
    completed:  '('
# STYLE
# color settings for session mode
"""

    if os.path.isfile(cfg_path):
        has_cfg = True
        with open(cfg_path, 'r') as fn:
            try:
                user = yaml.load(fn)
            except Exception as e:
                error = f"This exception was raised when loading settings:\n---\n{e}---\nPlease correct the error in {cfg_path} or remove it and restart nts.\n"
                logger.critical(error)
                sys.exit()
        if 'light_background' not in user:
            text = prompt(f"""\
d)ark or l)ight terminal background? [Dl] > """)
            if text.lower() == 'l':
                default_cfg = default_template + light
            else:
                default_cfg = default_template + dark
        elif user['light_background']:
            default_cfg = default_template + light
        else:
            default_cfg = default_template + dark
    else:
        has_cfg = False
        user = {}
        text = prompt(f"""\
Use color settings for a d)ark or l)ight terminal background? [Dl] > """)
        if text.lower() == 'l':
            default_cfg = default_template + light
        else:
            default_cfg = default_template + dark


    defaults = ruamel.yaml.load(default_cfg, ruamel.yaml.RoundTripLoader)
    merged = deepcopy(defaults)

    if has_cfg:
        changes = []
        for key, value in defaults.items():
            # if there is a user setting, use it - else use the default
            if key in user:
                if key == 'style':
                    for k, v in defaults['style'].items():
                        if user['style'] is not None and k in user['style']:
                            merged['style'][k] = user['style'][k]
                        else:
                            # a missing user setting component - use the default and update the file
                            changes.append(f"replaced missing setting for style[{k}] with {v}")

                    if user and 'style' in user and isinstance(user['style'], dict):
                        for k, v in user['style'].items():
                            if k not in merged['style']:
                                changes.append(f'removed invalid "{k}" setting for style')
                else:
                    merged[key] = user[key]
            else:
                # a missing user setting - use the default and update the file
                changes.append(f"replaced missing setting for {key} with {value}")

        for key, value in user.items():
            if key not in merged:
                changes.append(f'removed invalid "{key}" setting')
        if changes:
            with open(cfg_path, 'w', encoding='utf-8') as fn:
                yaml.dump(merged, fn)
            logger.info(f"updated {cfg_path}: {', '.join(changes)}")

    else:
        with open(cfg_path, 'w', encoding='utf-8') as fn:
            yaml.dump(merged, fn)

    with open(cfg_path, 'r') as fo:
        yaml_data = yaml.load(fo)

    return yaml_data


def make_grandchild(rootdir):
    grandchild = """\
+ note a (red, green)
    The body of note a goes here

+ note b (blue, green)
    The body of note b goes here

+ note c (red, blue)
    And the body of note c goes here
"""
    tagsort = """\
+ action required as soon as possible (now)
    In tag view, items with this tag will be sorted
    first

+ action needed when time permits (next)
    In tag view, items with this tag will be sorted
    after 'now' items

+ assigned for action (assigned joe)
    In tag view, items with this tag will be sorted
    in a third group and, within that group, by the
    name to whom it was assigned

+ assigned for action (assigned bob)
    In tag view, items with this tag will be sorted
    in a third group and, within that group, by the
    name to whom it was assigned

+ review from time to time for action (someday)
    In tag view, items with this tag will be sorted
    after all tagged items other than 'completed'

+ finished but kept for reference (completed)
    In tag view, items with this tag will be sorted
    after all tagged items

+ a note with no tags
    In tag view, such items will be sorted last under
    the implicit tag '~'
"""
    full_path = rootdir
    for path in ['parent', 'child']:
        full_path = os.path.join(full_path, path)
        if not os.path.isdir(full_path):
            os.mkdir(full_path)
    grandchild_path = os.path.join(full_path, 'grandchild.txt')
    with open(grandchild_path, 'w') as fo:
        fo.write(grandchild)
    print(f"created '{full_path}'")

    tagsort_path = os.path.join(rootdir, 'tagsort.txt')
    with open(tagsort_path, 'w') as fo:
        fo.write(tagsort)
    print(f"created '{full_path}'")

def setup_logging(level, ntsdir, file=None):
    """
    Setup logging configuration. Override root:level in
    logging.yaml with default_level.
    """

    if not os.path.isdir(ntsdir):
        return

    log_levels = {
        1: logging.DEBUG,
        2: logging.INFO,
        3: logging.WARN,
        4: logging.ERROR,
        5: logging.CRITICAL
    }

    level = int(level)
    loglevel = log_levels.get(level, log_levels[3])

    # if we get here, we have an existing ntsdir
    logfile = os.path.normpath(os.path.abspath(os.path.join(ntsdir, "nts.log")))

    config = {'disable_existing_loggers': False,
              'formatters': {'simple': {
                  'format': '--- %(asctime)s - %(levelname)s - %(module)s.%(funcName)s\n    %(message)s'}},
              'handlers': {
                    'file': {
                        'backupCount': 7,
                        'class': 'logging.handlers.TimedRotatingFileHandler',
                        'encoding': 'utf8',
                        'filename': logfile,
                        'formatter': 'simple',
                        'level': loglevel,
                        'when': 'midnight',
                        'interval': 1}
              },
              'loggers': {
                  'etmmv': {
                    'handlers': ['file'],
                    'level': loglevel,
                    'propagate': False}
              },
              'root': {
                  'handlers': ['file'],
                  'level': loglevel},
              'version': 1}
    logging.config.dictConfig(config)
    logger.critical("\n######## Initializing logging #########")
    if file:
        logger.critical(f'logging for file: {file}\n    logging at level: {loglevel}\n    logging to file: {logfile}')
    else:
        logger.critical(f'logging at level: {loglevel}\n    logging to file: {logfile}')


def main():
    import nts
    import nts.__version__ as version
    nts_version = version.version

    logging.getLogger('asyncio').setLevel(logging.WARNING)
    logger = logging.getLogger()
    MIN_PYTHON = (3, 7, 3)
    if sys.version_info < MIN_PYTHON:
        mv = ".".join([str(x) for x in MIN_PYTHON])
        sys.exit(f"Python {mv} or later is required.\n")
    import os
    IS_VENV = os.getenv('VIRTUAL_ENV') is not None
    import nts.__version__ as version

    cwd = os.getcwd()
    dlst = [x for x in os.listdir(cwd) if not x.startswith('.')]
    NTSHOME = os.environ.get("NTSHOME")
    if len(dlst) == 0 or ('data' in dlst and 'logs' in dlst) or ('cfg.yaml' in dlst and 'logs' in dlst):
        # use cwd if it is empty or contains both data and logs
        ntshome = cwd
    elif NTSHOME and os.path.isdir(NTSHOME):
        # else use NTSHOME if it is set and specifies a directory
        ntshome = NTSHOME
    else:
        # use the default ~/nts
        ntshome = os.path.join(os.path.expanduser('~'), 'nts')
    if not os.path.isdir(ntshome):
        text = prompt(f"'{ntshome}' does not exist. Create it [yN] > ")
        if text.lower().strip() == 'y':
            os.mkdir(ntsdir)
        else:
            print("cancelled")
            return

    logdir = os.path.normpath(os.path.join(ntshome, 'logs'))
    if not os.path.isdir(logdir):
        os.makedirs(logdir)
    loglevel = 2 # info
    log_levels = [str(x) for x in range(1, 6)]
    if len(sys.argv) > 1 and sys.argv[1] in log_levels:
        loglevel = int(sys.argv.pop(1))

    setup_logging(loglevel, logdir)
    logger.debug(f"nts home directory: '{ntshome}'")

    rootdir = os.path.join(ntshome, 'data')
    if not os.path.isdir(rootdir):
        os.makedirs(rootdir)
        logger.info(f"Created '{rootdir}'")
        text = prompt(f"populate {rootdir} with example data? [yN] > ")
        if text.lower().strip() == 'y':
            make_grandchild(rootdir)
            logger.info("added example data")

    cfg_path = os.path.join(ntshome, 'cfg.yaml')

    import nts.nts as nts
    nts.logger = logger
    Data = nts.NodeData(rootdir)
    nts.Data = Data
    nts.nts_version = nts_version
    nts.get_yaml_data = get_yaml_data
    nts.cfg_path = cfg_path

    yaml_data = get_yaml_data(cfg_path)
    if yaml_data:
        session_edit = f"{yaml_data['edit_command']} {yaml_data['session_edit_args']}"
        nts.session_edit = session_edit
        session_add = f"{yaml_data['edit_command']} {yaml_data['session_add_args']}"
        nts.session_add = session_add
        command_edit = f"{yaml_data['edit_command']} {yaml_data['command_edit_args']}"
        nts.command_edit = command_edit
        command_add = f"{yaml_data['edit_command']} {yaml_data['command_add_args']}"
        nts.command_add = command_add
        user_style = yaml_data['style']
        style_obj = Style.from_dict(user_style)
        nts.style_obj = style_obj
        tag_sort = yaml_data.get('tag_sort', {})
        nts.tag_sort = tag_sort

    nts.main()